chain/utils/src/crypto.rs

```diff
@@ -0,0 +1,428 @@
+        let mut bytes = [0u8; X25519_LEN];
+        value.read(&mut bytes)?;
+        Ok(Self::from(bytes))
+    }
+}
+
+#[derive(Default)]
+pub struct KeyChain {
+    keys: HashMap<u8, [u8; X25519_LEN]>,
+    public: HashMap<u8, HashSet<[u8; X25519_LEN]>>,
+}
+
+impl KeyChain {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn insert<T: KeyType>(&mut self, key: PrivateKey<T>) {
+        let public = PublicKey::from(&key);
+        self.keys.insert(T::KEY_TYPE, key.to_bytes());
+        let group = self.public.entry(T::KEY_TYPE).or_default();
+        group.insert(*public.as_bytes());
+    }
+
+    pub fn get<T: KeyType>(&self) -> Option<PrivateKey<T>> {
+        self.keys
+            .get(&T::KEY_TYPE)
+            .map(|bytes| PrivateKey::from(*bytes))
+    }
+
+    pub fn insert_public<T: KeyType>(&mut self, public: PublicKey<T>) {
+        let group = self.public.entry(T::KEY_TYPE).or_default();
+        group.insert(*public.as_bytes());
+    }
+
+    pub fn get_public<T: KeyType>(&self) -> Vec<PublicKey<T>> {
+        if let Some(set) = self.public.get(&T::KEY_TYPE) {
+            set.iter().map(|bytes| PublicKey::from(*bytes)).collect()
+        } else {
+            Default::default()
+        }
+    }
+}
+
+#[derive(Eq, PartialEq)]
+pub struct SecretBox<K, T> {
+    _marker: PhantomData<(K, T)>,
+    secret: Vec<u8>,
+}
+
+impl<K, T> Clone for SecretBox<K, T> {
+    fn clone(&self) -> Self {
+        Self {
+            _marker: self._marker,
+            secret: self.secret.clone(),
+        }
+    }
+}
+
+impl<K, T> std::fmt::Debug for SecretBox<K, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "SecretBox")
+    }
+}
+
+impl<K, T> Encode for SecretBox<K, T> {
+    fn size_hint(&self) -> usize {
+        self.secret.len()
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.secret.using_encoded(f)
+    }
+}
+
+impl<K, T> Decode for SecretBox<K, T> {
+    fn decode<R: Input>(value: &mut R) -> Result<Self, parity_scale_codec::Error> {
+        Ok(Self {
+            _marker: PhantomData,
+            secret: Decode::decode(value)?,
+        })
+    }
+}
+
+impl<K: KeyType, T: Decode + Encode> SecretBox<K, T> {
+    pub fn encrypt(key_chain: &KeyChain, payload: &T) -> Result<Self, SecretBoxError> {
+        let recipients = key_chain.get_public::<K>();
+        Self::encrypt_for(payload, &recipients)
+    }
+
+    pub fn encrypt_for(payload: &T, recipients: &[PublicKey<K>]) -> Result<Self, SecretBoxError> {
+        if recipients.is_empty() {
+            return Err(SecretBoxError::NoRecipients);
+        }
+        if recipients.len() as u8 as usize != recipients.len() {
+            return Err(SecretBoxError::TooManyRecipients);
+        }
+        // Create a buffer.
+        let capacity =
+            recipients.len() * (X25519_LEN + TAG_LEN) + X25519_LEN + 1 + payload.size_hint();
```

Comment:
No tag right? Because that goes into a content addressed block.
```suggestion
            recipients.len() * X25519_LEN + X25519_LEN + 1 + payload.size_hint();
```